docs/RFCS/20190731_cluster_name.md

```diff
@@ -0,0 +1,118 @@
+- Feature Name: cluster name
+- Status: draft
+- Start Date: 2019-07-31
+- Authors: knz, ben, marc
+- RFC PR: [#39196](https://github.com/cockroachdb/cockroach/pull/39196)
+- Cockroach Issue: [#16784](https://github.com/cockroachdb/cockroach/issues/16784) [#15888](https://github.com/cockroachdb/cockroach/issues/15888) [#28408](https://github.com/cockroachdb/cockroach/issues/28408)
+
+# Summary
+
+New feature: a string value, called "cluster name", and checked for
+equality when a newly started node joins a cluster.
+
+This will prevent newly added nodes to join the wrong cluster when a
+user has multiple clusters running side by side.
+
+This would be configured using another command-line parameter `--cluster-name`
+alongside `--join`.
+
+It will increase operational simplicity and the overall goal to “make
+data easy”.
+
+Other impact:
+
+- (should have) it would be checked for equality in the node
+  certificate properties. This would prevent a node from trusting
+  another node's certificate if not in the same cluster, even though
+  they may be signed by the same CA.
+
+- (nice to have) the value would be shown in the admin UI, so that two UI
+  screens for different clusters side-by-side can be disambiguated at
+  a glance.
+
+- (to be further considered) the value would be used as an additional
+  label annotation on exported prometheus metrics. This would make it
+  possible for Grafana and other monitoring solutions to more easily
+  connect to multiple CockroachDB clusters simultaneously.
+
+# Motivation
+
+The motivation for doing this has been detail in
+https://github.com/cockroachdb/cockroach/issues/16784 and
+https://github.com/cockroachdb/cockroach/issues/15888
+
+- avoid mistaken node joins
+- disambiguate UI screens
+- disambiguate clusters in prometheus metrics
+
+# Guide-level explanation
+
+The service manager in charge of starting the `cockroach` process
+would activate this feature by providing the `--cluster-name` parameter
+on the command line.
+
+When the parameter is supplied, the node will verify that the other
+nodes that it connects to (via `--join`) have the same name
+configured. If they detect a different name, the join would abort.
+
+If the parameter is not supplied, we get the current behavior - the
+join succeeds in any case.
+
+# Reference-level explanation
+
+- The parameter is parsed for `cockroach start` only.
+
+- It is stored in the `server.Config` object.
+
+- When setting up TLS, it is used to verify client certs (both nodes
+  verifying clients and nodes verifying other nodes). Nodes verifying
+  node server certs might be doable, but probably isn't needed.
+
+- When a newly created node (cluster ID not known yet) connects, it is
+  used to verify the identity of peers instead of the cluster
+  ID. (note: we already validate the cluster ID beyond that).
+
+- (Optionally) reported in `crdb_internal.gossip_nodes` for troubleshooting.
```

Comment:
A sql builtin to return the cluster name would be convenient as well.

Reply:
Done.